### PR TITLE
Make action buttons sticky on mobile widgets

### PR DESCRIPTION
**Pull request by Larisa-Staroverova**

### CODE REVIEW

File: feature-libs/quote/core/services/quote-storefront-utils.service.ts [hunk 1]
```diff
@@ -26,4 +26,75 @@ export class QuoteStorefrontUtilsService {
...
+    return false;
+  }
+
+  /**
+   * Retrieves the height of the HTML element.
+   *
+   * @param querySelector - querySelector
+   * @returns - the height of the HTML element
+   */
+  public getHeight(querySelector: string): number {
```
Comment (ChristophHi) [inline feature-libs/quote/core/services/quote-storefront-utils.service.ts]: Please remove 'public'

File: feature-libs/quote/core/services/quote-storefront-utils.service.ts [hunk 2]
```diff
@@ -26,4 +26,75 @@ export class QuoteStorefrontUtilsService {
...
+    const element = this.getElement(querySelector);
+    const isElementInViewport = this.isInViewport(element);
+    if (isElementInViewport && element?.offsetHeight) {
+      return element?.offsetHeight;
+    }
+    return 0;
+  }
+
+  public getWindowHeight() {
```
Comment (ChristophHi) [inline feature-libs/quote/core/services/quote-storefront-utils.service.ts]: Please remove 'public'

File: feature-libs/quote/core/services/quote-storefront-utils.service.ts [hunk 3]
```diff
@@ -26,4 +26,75 @@ export class QuoteStorefrontUtilsService {
...
+    if (this.windowRef.isBrowser()) {
+      return this.windowRef.nativeWindow
+        ? this.windowRef.nativeWindow.innerHeight
+        : 0;
```
Comment (ChristophHi) [inline feature-libs/quote/core/services/quote-storefront-utils.service.ts:123]: Not covered, can we also test that?

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 4]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
       ).toBe(QuoteRoleType.ALL);
     });
   });
+
+  describe('isMobile', () => {
+    it('should not render in mobile mode', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]:
```suggestion
    it('should return false in mobile mode', () => {
```

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 5]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+      spyOn(breakpointService, 'isDown').and.returnValue(of(false));
+      fixture.detectChanges();
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: Please delete, not needed

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 6]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+
+      component['isMobile']().subscribe((isMobile) => {
+        expect(isMobile).toBe(false);
+      });
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]:
```suggestion
      }).unsubscribe();
```

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 7]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+    });
+
+    it('should render in mobile mode', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: Please see all comments for previous method

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 8]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+      fixture.detectChanges();
+
+      component['isMobile']().subscribe((isMobile) => {
+        expect(isMobile).toBe(true);
+      });
+    });
+  });
+
+  describe('isDesktop', () => {
+    it('should not render in desktop mode', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: Please see comments for previous method

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 9]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+      spyOn(breakpointService, 'isUp').and.returnValue(of(false));
+      fixture.detectChanges();
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: remove

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 10]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+
+      component['isDesktop']().subscribe((isMobile) => {
+        expect(isMobile).toBe(false);
+      });
+    });
+
+    it('should render in desktop mode', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: Please see comments for previous method (does not really relate to 'render', also detectChanges not needed etc.

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts [hunk 11]
```diff
@@ -71,6 +137,177 @@ export class QuoteSummaryActionsComponent implements OnInit, OnDestroy {
...
+  /**
+   * Prepares the styling of the action buttons to be displayed on mobile.
+   *
+   * @protected
+   */
+  protected prepareButtonsForMobile(): void {
+    this.isMobile()
+      .pipe(take(1))
+      .subscribe((mobile) => {
+        if (mobile) {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: Please add a 'filter' instead of the if statement, in that case we don't need to subscribe (and expression is more concise). Do you agree?

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts [hunk 12]
```diff
@@ -71,6 +137,177 @@ export class QuoteSummaryActionsComponent implements OnInit, OnDestroy {
...
+  /**
+   * Prepares the styling of the action buttons to be displayed on desktop.
+   *
+   * @protected
+   */
+  protected prepareButtonsForDesktop(): void {
+    this.isDesktop()
+      .pipe(take(1))
+      .subscribe((desktop) => {
+        if (desktop) {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: Please add a 'filter' instead of the if statement, in that case we don't need to subscribe (and expression is more concise). Do you agree?

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts [hunk 13]
```diff
@@ -71,6 +137,177 @@ export class QuoteSummaryActionsComponent implements OnInit, OnDestroy {
     });
   }
 
+  protected isDesktop() {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: Is it either desktop or mobile? Then we could have _one_ method instead of 2. Does that make sense?

Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: Please add return type

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 14]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+        expect(isMobile).toBe(true);
+      });
+    });
+  });
+
+  describe('handleResize', () => {
+    it('should call handleResize method', (done) => {
+      component.handleResize();
+
+      component.quoteDetails$.pipe(take(1), delay(0)).subscribe(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: Why is that needed?
This test also passes
```

    it('should call handleResize method', () => {
      component.handleResize();

      expect(quoteStorefrontUtilsService.changeStyling).toHaveBeenCalledWith(
        'cx-quote-summary-actions section',
        'position',
        'static'
      );
    });
```

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 15]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+          'static'
+        );
+        done();
+      });
+    });
+  });
+
+  describe('handleScroll', () => {
+    it('should call handleScroll method', (done) => {
+      spyOn(quoteStorefrontUtilsService, 'getWindowHeight').and.returnValue(
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts:972]: Why not like this:
 ```
   it('should call handleScroll method', () => {
      spyOn(quoteStorefrontUtilsService, 'getWindowHeight').and.returnValue(
        500
      );
      spyOn(breakpointService, 'isDown').and.returnValue(of(true));
      component.handleScroll();

      expect(quoteStorefrontUtilsService.changeStyling).toHaveBeenCalledWith(
        'cx-quote-summary-actions section',
        'bottom',
        '0'
      );
```

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 16]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+      expect(component['getActionButtonsHeight']()).toBe(226);
+    });
+
+    it('should return the actual height of action buttons', () => {
+      spyOn(quoteStorefrontUtilsService, 'getHeight').and.returnValue(300);
+      expect(component['getActionButtonsHeight']()).toBe(300);
+    });
+  });
+
+  describe('Floating action buttons', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts:998]: Why not like this

 ```
   it('should make action buttons static for desktop', () => {
      component.ngAfterViewInit();

      expect(quoteStorefrontUtilsService.changeStyling).toHaveBeenCalledWith(
        'cx-quote-summary-actions section',
        'position',
        'static'
      );
    });
```

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 17]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+        .and.returnValue(130)
+        .withArgs('cx-quote-summary-actions section')
+        .and.returnValue(250);
+
+      spyOn(quoteStorefrontUtilsService, 'getWindowHeight').and.returnValue(
+        500
+      );
+      spyOn(breakpointService, 'isDown').and.returnValue(of(true));
+      component.ngAfterViewInit();
+      component.quoteDetails$.pipe(take(1), delay(0)).subscribe(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: See previous comment

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 18]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+        .and.returnValue(130)
+        .withArgs('cx-quote-summary-actions section')
+        .and.returnValue(250);
+
+      spyOn(quoteStorefrontUtilsService, 'getWindowHeight').and.returnValue(
+        300
+      );
+      spyOn(breakpointService, 'isDown').and.returnValue(of(true));
+      component.ngAfterViewInit();
+      component.quoteDetails$.pipe(take(1), delay(0)).subscribe(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: See previous comment

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 19]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+        );
+        done();
+      });
+    });
+
+    it('should make action buttons sticky when intersecting', (done) => {
+      spyOn(intersectionService, 'isIntersecting').and.returnValue(of(true));
+      spyOn(breakpointService, 'isDown').and.returnValue(of(true));
+      component.ngAfterViewInit();
+      component.quoteDetails$.pipe(take(1), delay(0)).subscribe(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: See previous comment

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts [hunk 20]
```diff
@@ -723,4 +776,195 @@ describe('QuoteSummaryActionsComponent', () => {
...
+          'sticky'
+        );
+        done();
+      });
+    });
+
+    it('should make action buttons fixed when not intersecting', (done) => {
+      spyOn(breakpointService, 'isDown').and.returnValue(of(true));
+      component.ngAfterViewInit();
+      component.quoteDetails$.pipe(take(1), delay(0)).subscribe(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.spec.ts]: See previous comment

File: feature-libs/quote/core/services/quote-storefront-utils.service.spec.ts [hunk 21]
```diff
@@ -1,61 +1,226 @@
...
+  let querySelectorOriginal: any;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      declarations: [MockQuoteComponent],
+      schemas: [CUSTOM_ELEMENTS_SCHEMA],
+    }).compileComponents();
+  });
+
+  beforeEach(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/core/services/quote-storefront-utils.service.spec.ts]: Please unify the 2 beforeEach methods

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts (hunk 13 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: Agree!

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts (hunk 12 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: Agree!

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts [hunk 22]
```diff
@@ -71,6 +137,169 @@ export class QuoteSummaryActionsComponent implements OnInit, OnDestroy {
...
+
+  /**
+   * Prepares the styling of the action buttons to be displayed on desktop.
+   *
+   * @protected
+   */
+  protected prepareButtonsForDesktop(): void {
+    this.isMobile()
+      .pipe(filter((mobile) => !mobile))
+      .subscribe(() => {
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts:234]: take(1) is missing

File: feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts [hunk 23]
```diff
@@ -71,6 +137,169 @@ export class QuoteSummaryActionsComponent implements OnInit, OnDestroy {
...
+  }
+
+  /**
+   * Prepares the styling of the action buttons to be displayed on mobile.
+   *
+   * @protected
+   */
+  protected prepareButtonsForMobile(): void {
+    this.isMobile()
+      .pipe(filter((mobile) => mobile))
```
Comment (ChristophHi) [inline feature-libs/quote/components/summary/actions/quote-summary-actions.component.ts]: take(1) is missing

**Review by ChristophHi — CHANGES_REQUESTED** [review]

**Review by ChristophHi — APPROVED** [review]

File: feature-libs/quote/core/services/quote-storefront-utils.service.spec.ts [hunk 24]
```diff
@@ -210,6 +210,26 @@ describe('QuoteStorefrontUtilsService', () => {
     });
   });
 
+  describe('getDomRectValue', () => {
+    it('should return undefined because no element is found by a selector query', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/core/services/quote-storefront-utils.service.spec.ts]: should return undefined if no element is found by a selector query

File: feature-libs/quote/core/services/quote-storefront-utils.service.spec.ts [hunk 25]
```diff
@@ -210,6 +210,26 @@ describe('QuoteStorefrontUtilsService', () => {
...
+      expect(
+        classUnderTest['getDomRectValue']('unknown-query', 'bottom')
+      ).toBeUndefined();
+    });
+
+    it('should return undefined because element does not contain a searched property', () => {
```
Comment (ChristophHi) [inline feature-libs/quote/core/services/quote-storefront-utils.service.spec.ts]: should return undefined if element does not contain a searched property

**Review by ChristophHi — CHANGES_REQUESTED** [review]

**Review by ChristophHi — APPROVED** [review]
